### PR TITLE
Added local testing support for S3

### PR DESCRIPTION
**Pull request by raphkim**
*Issue #, if available:*
N/A 

*Description of changes:*
Added local testing support for S3. It reads `DangerouslyConnectToHTTPEndpointForTesting` flag from awsconfiguration.json file to determine if a http client should be configured to connect with local server.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws-android-sdk-s3/src/main/java/com/amazonaws/mobileconnectors/s3/transferutility/TransferUtility.java [hunk 1]
```diff
@@ -264,6 +266,20 @@ public TransferUtility build() {
...
+                    // Checks if awsconfiguration.json has local testing flag to dangerously connect to HTTP endpoint.
+                    // Defaults to false unless specified.
+                    final boolean canConnectToHTTPEndpoint = tuConfig.has(Constants.LOCAL_TESTING_FLAG_NAME) ?
+                            tuConfig.getBoolean(Constants.LOCAL_TESTING_FLAG_NAME) : false;
+
+                    // Mutates AmazonS3Client object to have local endpoint
+                    if (canConnectToHTTPEndpoint) {
+                        this.s3.setEndpoint(Constants.LOCAL_TESTING_ENDPOINT);
+                        this.s3.setS3ClientOptions(S3ClientOptions.builder()
+                                .setPathStyleAccess(true)
```
Comment (kvasukib) [inline aws-android-sdk-s3/src/main/java/com/amazonaws/mobileconnectors/s3/transferutility/TransferUtility.java]: Add a comment for each of these option as to why we are setting these flags to true.

File: aws-android-sdk-s3/src/main/java/com/amazonaws/services/s3/S3ClientOptions.java [hunk 2]
```diff
@@ -305,6 +352,22 @@ public boolean isPayloadSigningEnabled() {
...
+     * </p>
+     * <p>
+     * Setting this flag will allow the client to disregard data transfer
+     * integrity check by bypassing content MD5 check for all requests and
+     * responses.
+     * </p>
+     *
+     * @param skipContentMd5Check True to always skip content MD5 check
+     */
+    public void setSkipContentMd5Check(boolean skipContentMd5Check) {
```
Comment (kvasukib) [inline aws-android-sdk-s3/src/main/java/com/amazonaws/services/s3/S3ClientOptions.java]: Suggestion to rename to: `skipContentMd5Check`

File: aws-android-sdk-s3/src/main/java/com/amazonaws/services/s3/S3ClientOptions.java [hunk 3]
```diff
@@ -200,25 +227,45 @@ public S3ClientOptions() {
...
+     * </p>
+     * <p>
+     * Setting this flag will allow the client to disregard data transfer
+     * integrity check by bypassing content MD5 check for all requests and
+     * responses.
+     * </p>
+     *
+     * @return True if the client should always skip content MD5 check
+     */
+    public boolean shouldSkipContentMd5Check() {
```
Comment (kvasukib) [inline aws-android-sdk-s3/src/main/java/com/amazonaws/services/s3/S3ClientOptions.java]: isContentMd5CheckSkipped()

File: aws-android-sdk-s3/src/main/java/com/amazonaws/services/s3/internal/Constants.java [hunk 4]
```diff
@@ -39,6 +39,11 @@
     /** Service hostname for accessing dualstack accelerated S3 buckets */
     public static final String S3_ACCELERATE_DUALSTACK_HOSTNAME = "s3-accelerate.dualstack.amazonaws.com";
 
+    /** Parameter name to indicate permission to dangerously connect to HTTP endpoint */
+    public static final String LOCAL_TESTING_FLAG_NAME = "DangerouslyConnectToHTTPEndpointForTesting";
```
Comment (kvasukib) [inline aws-android-sdk-s3/src/main/java/com/amazonaws/services/s3/internal/Constants.java]: Add a newline after this.

File: aws-android-sdk-s3/src/main/java/com/amazonaws/services/s3/internal/Constants.java [hunk 5]
```diff
@@ -39,6 +39,11 @@
...
+    /** Default endpoint for the local S3 storage testing */
+    public static final String LOCAL_TESTING_ENDPOINT = "http://10.0.2.2:20005";
```
Comment (kvasukib) [inline aws-android-sdk-s3/src/main/java/com/amazonaws/services/s3/internal/Constants.java]: Add a comment here that this IP is used by Android emulators to connect to localhost.

**Comment by raphkim** [comment]
I'm considering switching AmazonS3Client's client option access modifier again from

`S3ClientOptions clientOptions = new S3ClientOptions();`

to

`protected S3ClientOptions clientOptions = new S3ClientOptions();`

I know this gives it unnecessarily broader access, but I think it makes sense to keep this instance variable accessible to its subclasses since `setS3ClientOptions()` is required by the AmazonS3 interface anyways.

Also, thoughts on re-adding `skipMd5CheckPerRequest(AmazonWebServiceRequest request)` and `skipMd5CheckPerResponse(ObjectMetadata metadata)` back, and then overloading the methods with `skipMd5CheckPerRequest(AmazonWebServiceRequest request, S3ClientOptions clientOptions)` and `skipMd5CheckPerResponse(ObjectMetadata metadata, S3ClientOptions clientOptions)`?

This is completely unnecessary, but this would at least remove the need to change `TransferManager.java` file just to add that additional parameter as well as portions of unit tests.

**Comment by kvasukib** [comment]
@raphkim Agreed with your changes. Please proceed. I have few minor comments on the rest of the code which I have posted.

File: aws-android-sdk-s3/src/main/java/com/amazonaws/services/s3/S3ClientOptions.java (hunk 2 above)
Comment (raphkim) [reply aws-android-sdk-s3/src/main/java/com/amazonaws/services/s3/S3ClientOptions.java]: this overlaps with the flag's variable name unfortunately

Comment (raphkim) [reply aws-android-sdk-s3/src/main/java/com/amazonaws/services/s3/S3ClientOptions.java]: didn't matter. will change.

File: aws-android-sdk-s3/src/main/java/com/amazonaws/mobileconnectors/s3/transferutility/TransferUtility.java (hunk 1 above)
Comment (raphkim) [reply aws-android-sdk-s3/src/main/java/com/amazonaws/mobileconnectors/s3/transferutility/TransferUtility.java]: will do

File: aws-android-sdk-s3/src/main/java/com/amazonaws/services/s3/S3ClientOptions.java (hunk 3 above)
Comment (raphkim) [reply aws-android-sdk-s3/src/main/java/com/amazonaws/services/s3/S3ClientOptions.java]: will rename this and others

**Pull request merged**